server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java

```diff
@@ -470,6 +473,26 @@ protected void updateIndicesReadOnly(Set<String> indicesToUpdate, ActionListener
             .execute(wrappedListener.map(r -> null));
     }
 
+    protected void updateIndicesRemoveReadOnlyBlock(boolean enabled) {
+        if (enabled) {
+            return;
+        }
+        final ClusterState state = clusterStateSupplier.get();
+        final Set<String> indicesToRelease = state.routingTable()
+            .indicesRouting()
+            .keySet()
+            .stream()
+            .filter(index -> state.getBlocks().hasIndexBlock(index, IndexMetadata.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK))
+            .collect(Collectors.toSet());
+        logger.info("removing read-only block from indices [{}]", indicesToRelease);
+        client.admin()
```

Comment:
Please check 35aa31230913e44d80eb44cac284f227636cd38a. (There has been some name changes, etc, as the PR was in draft, and I wanted to clean it up).

Reply:
Unfortunately I don't think it's quite enough to make these methods `synchronized` - `updateIndicesReadOnly` is an async method which returns, releasing the mutex, before it has done anything.

Having said that, it's possible that this does fix the race because of how the client's `execute()` method at least enqueue the update before returning, but I think that's a subtle implementation detail on which it's not safe to rely. Could we instead use the `checkInProgress` flag to more explicitly avoid concurrency here?

Reply:
I see! Good point. I guess in that case, we don't need `synchronized` at all. But if we are considering the order of the async requests not to be enough here, then I guess the following could also be a problem: 

Setting is disabled -> `removeExistingIndexBlocksIfDisabled` is called and we queue an async request to remove the existing blocks. 

While this is in-flight, setting gets re-enabled , an `onNewInfo` call follows at some point after that. Now the `onNewInfo` (`updateIndicesReadOnly`) call needs to wait for the request to remove blocks, as otherwise, similar to the case you mentioned, the blocks that it could potentially put on indices could be removed by the request sent from `removeExistingIndexBlocksIfDisabled`.

I think, we don't need the synchronized anymore, but for the case you mentioned, we could stall the`removeExistingIndexBlocksIfDisabled` call until there is no more checks in progress, but then we'd need also another "disable in progress" to stall/reject `onNewInfo` calls. Is that the case?

Reply:
I think this is not so bad since `onNewInfo` is called repeatedly so eventually the blocks will be added back, but in any case I think it would be enough to use the existing `checkInProgress` to fix this.

Reply:
> I think it would be enough to use the existing `checkInProgress` to fix this.

I'm not sure about this now that I think about it a bit more. We might need a bit more structure here.

Reply:
> I think this is not so bad since onNewInfo is called repeatedly so eventually the blocks will be added back

Then why not just rely on the periodic `onNewInfo` calls to handle disabling the setting? If it is disabled at the end of `onNewInfo`, I'd just make a call to remove existing blocks if there are any. This means from the 3 callbacks that the check is waiting for, one would be either `updateIndicesReadOnly` or `removeExistingIndexBlocksIfDisabled` (the name should be changed then), depending on whether the setting is enabled. WDYT?

Downside is that disabling the setting would not take effect immediately.

Reply:
I've tried to do it with `checkInProgress` in https://github.com/elastic/elasticsearch/pull/87841/commits/54aa7687d99353771c747d990ede7bc2e7e5e4d7. Do you see a case, where this wouldn't work?

Reply:
@elasticmachine update branch

Reply:
> I'm not sure about this now that I think about it a bit more. We might need a bit more structure here.

@DaveCTurner You mentioned there might be an issue if we only rely on `checkInProgress`. Could you please elaborate on that a bit?